File: src/python/offer/22.链表中倒数第k个节点.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   22.链表中倒数第k个节点.py
@Time    :   2020/08/24 19:30:27
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def getKthFromEnd(self, head: ListNode, k: int) -> ListNode:
        if not head or k <= 0:
            return

        # 1->2->3->4->5  k = 2
        slow, fast = head, head
        for _ in range(k):
            fast = fast.next

        while fast:
            slow = slow.next
            fast = fast.next

        return slow

File: src/golang/offer/22.链表中倒数第k个节点.go
package offer

/**
 * @File    :   22.链表中倒数第k个节点.go
 * @Time    :   2021/09/02 09:37:49
 * @Author  :   wylu
 * @Version :   1.0
 * @Contact :   15wylu@gmail.com
 * @License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
 * @Desc    :   https://leetcode-cn.com/problems/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof/
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getKthFromEnd(head *ListNode, k int) *ListNode {
	slow, fast := head, head
	for ; k > 0; k-- {
		fast = fast.Next
	}

	for fast != nil {
		slow = slow.Next
		fast = fast.Next
	}

	return slow
}
